File: src/macos/LanMessengerTests/VideoFrameBudgetTests.swift
import XCTest
@testable import LanMessenger

/// The protocol's two-frame rule, which PROTOCOL.md states once and five
/// independent places have to keep. Mirror of `VideoFrameBudgetTests.cs`.
final class VideoFrameBudgetTests: XCTestCase {

    func testTheThirdFrameIsRefused() {
        // One being worked on, one waiting. Fixed by the protocol, not tuned.
        let budget = VideoFrameBudget()
        XCTAssertTrue(budget.tryAcquire())
        XCTAssertTrue(budget.tryAcquire())
        XCTAssertFalse(budget.tryAcquire())
        XCTAssertEqual(budget.inFlight, 2)
        XCTAssertEqual(budget.refused, 1)
    }

    func testARefusalIsCountedRatherThanSwallowed() {
        // A budget that is constantly full is a real signal about the encoder.
        // Dropping silently is how that signal is lost — and "it feels laggy"
        // becomes the only available measurement again.
        let budget = VideoFrameBudget()
        _ = budget.tryAcquire(); _ = budget.tryAcquire()
        for _ in 0..<5 { XCTAssertFalse(budget.tryAcquire()) }
        XCTAssertEqual(budget.refused, 5)
        XCTAssertTrue(budget.summary.contains("refused=5"))
    }

    func testAReleasedSlotComesBack() {
        let budget = VideoFrameBudget()
        _ = budget.tryAcquire(); _ = budget.tryAcquire()
        budget.release()
        XCTAssertTrue(budget.tryAcquire())
        XCTAssertEqual(budget.inFlight, 2)
    }

    func testReleasingMoreThanWasTakenCannotWidenTheBudget() {
        // An encoder that emits two outputs for one input would otherwise drive
        // the count negative, and the budget would stop bounding anything at
        // all — a slow drift back into unbounded latency with nothing to see.
        let budget = VideoFrameBudget()
        _ = budget.tryAcquire()
        budget.release()
        budget.release()
        budget.release()
        XCTAssertEqual(budget.inFlight, 0)

        XCTAssertTrue(budget.tryAcquire())
        XCTAssertTrue(budget.tryAcquire())
        XCTAssertFalse(budget.tryAcquire(), "the cap held after an over-release")
    }

    func testResetForgetsWhatWillNeverComeBack() {
        // Flush, restart and teardown. A slot leaked across one of those shrinks
        // the budget for the rest of the session, and after two the encoder
        // accepts nothing with no error anywhere to explain it.
        let budget = VideoFrameBudget()
        _ = budget.tryAcquire(); _ = budget.tryAcquire()
        budget.reset()
        XCTAssertEqual(budget.inFlight, 0)
        XCTAssertTrue(budget.tryAcquire())
    }

    func testTheCapacityIsTheProtocolsNumber() {
        // Not a tuning knob. PROTOCOL.md fixes it, and both platforms say 2.
        XCTAssertEqual(VideoFrameBudget.protocolCapacity, 2)
        XCTAssertEqual(VideoFrameBudget().capacity, 2)
    }

    func testACodecPipelineGetsARunawayGuardRatherThanTheQueueRule() {
        // The mistake this encodes: two is right for a queue and wrong for a
        // hardware transform. Quick Sync issues one METransformNeedInput per
        // pipeline slot and emits nothing until enough are filled — capped at
        // two it produced no video at all, and not one encoder_stats line to
        // say why. A pipeline's depth is fixed latency, not growth.
        XCTAssertGreaterThan(VideoFrameBudget.pipelineCapacity,
                             VideoFrameBudget.protocolCapacity)

        let pipeline = VideoFrameBudget(capacity: VideoFrameBudget.pipelineCapacity)
        for _ in 0..<VideoFrameBudget.pipelineCapacity {
            XCTAssertTrue(pipeline.tryAcquire(), "a pipeline must not be starved")
        }
        XCTAssertFalse(pipeline.tryAcquire(), "but a runaway is still caught")
    }

    func testThePeakIsReportedBecauseNothingElseCanSeeIt() {
        // The codec's own depth is latency no other measurement reaches, and
        // exposing it is the point of the encoder's budget now that it is not
        // pacing anything.
        let budget = VideoFrameBudget(capacity: 8)
        for _ in 0..<5 { _ = budget.tryAcquire() }
        for _ in 0..<5 { budget.release() }

        XCTAssertEqual(budget.inFlight, 0)
        XCTAssertEqual(budget.highWater, 5)
        XCTAssertTrue(budget.summary.contains("peak=5"), budget.summary)
    }

    func testConcurrentAcquiresNeverExceedTheCap() {
        // The capture thread submits and the encoder's callback releases, on
        // different threads. A cap that only holds single-threaded is not a cap.
        let budget = VideoFrameBudget()
        let done = expectation(description: "all attempts finished")
        done.expectedFulfillmentCount = 8

        let peak = NSLock()
        var highWater = 0

        for _ in 0..<8 {
            DispatchQueue.global().async {
                for _ in 0..<500 {
                    if budget.tryAcquire() {
                        peak.lock()
                        highWater = max(highWater, budget.inFlight)
                        peak.unlock()
                        budget.release()
                    }
                }
                done.fulfill()
            }
        }

        wait(for: [done], timeout: 20)
        XCTAssertLessThanOrEqual(highWater, VideoFrameBudget.protocolCapacity)
        XCTAssertEqual(budget.inFlight, 0)
    }
}

File: src/windows-native/LanMessenger.Tests/VideoFrameBudgetTests.cs
using LanMessenger.Core.Networking.Media;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading.Tasks;

namespace LanMessenger.Tests;

/// <summary>
/// The protocol's two-frame rule, which PROTOCOL.md states once and five
/// independent places have to keep. Mirror of VideoFrameBudgetTests.swift.
/// </summary>
[TestClass]
public class VideoFrameBudgetTests
{
    [TestMethod]
    public void TheThirdFrameIsRefused()
    {
        // One being worked on, one waiting. Fixed by the protocol, not tuned.
        var budget = new VideoFrameBudget();
        Assert.IsTrue(budget.TryAcquire());
        Assert.IsTrue(budget.TryAcquire());
        Assert.IsFalse(budget.TryAcquire());
        Assert.AreEqual(2, budget.InFlight);
        Assert.AreEqual(1, budget.Refused);
    }

    [TestMethod]
    public void ARefusalIsCountedRatherThanSwallowed()
    {
        // A budget that is constantly full is a real signal about the encoder.
        // Dropping silently is how that signal is lost — and "it feels laggy"
        // becomes the only available measurement again.
        var budget = new VideoFrameBudget();
        budget.TryAcquire(); budget.TryAcquire();
        for (int i = 0; i < 5; i++) Assert.IsFalse(budget.TryAcquire());
        Assert.AreEqual(5, budget.Refused);
        StringAssert.Contains(budget.Summary(), "refused=5");
    }

    [TestMethod]
    public void AReleasedSlotComesBack()
    {
        var budget = new VideoFrameBudget();
        budget.TryAcquire(); budget.TryAcquire();
        budget.Release();
        Assert.IsTrue(budget.TryAcquire());
        Assert.AreEqual(2, budget.InFlight);
    }

    [TestMethod]
    public void ReleasingMoreThanWasTakenCannotWidenTheBudget()
    {
        // An encoder that emits two outputs for one input would otherwise drive
        // the count negative, and the budget would stop bounding anything at
        // all — a slow drift back into unbounded latency with nothing to see.
        var budget = new VideoFrameBudget();
        budget.TryAcquire();
        budget.Release();
        budget.Release();
        budget.Release();
        Assert.AreEqual(0, budget.InFlight);

        Assert.IsTrue(budget.TryAcquire());
        Assert.IsTrue(budget.TryAcquire());
        Assert.IsFalse(budget.TryAcquire(), "the cap held after an over-release");
    }

    [TestMethod]
    public void ResetForgetsWhatWillNeverComeBack()
    {
        // Flush, drain-complete and teardown. A slot leaked across one of those
        // shrinks the budget for the rest of the session, and after two the
        // encoder accepts nothing with no error anywhere to explain it.
        var budget = new VideoFrameBudget();
        budget.TryAcquire(); budget.TryAcquire();
        budget.Reset();
        Assert.AreEqual(0, budget.InFlight);
        Assert.IsTrue(budget.TryAcquire());
    }

    [TestMethod]
    public void TheCapacityIsTheProtocolsNumber()
    {
        // Not a tuning knob. PROTOCOL.md fixes it, and both platforms say 2.
        Assert.AreEqual(2, VideoFrameBudget.ProtocolCapacity);
        Assert.AreEqual(2, new VideoFrameBudget().Capacity);
    }

    [TestMethod]
    public void ACodecPipelineGetsARunawayGuardRatherThanTheQueueRule()
    {
        // The mistake this encodes: two is right for a queue and wrong for a
        // hardware transform. Quick Sync issues one METransformNeedInput per
        // pipeline slot and emits nothing until enough are filled — capped at
        // two it produced no video at all, and not one encoder_stats line to
        // say why. A pipeline's depth is fixed latency, not growth.
        Assert.IsTrue(VideoFrameBudget.PipelineCapacity > VideoFrameBudget.ProtocolCapacity);

        var pipeline = new VideoFrameBudget(VideoFrameBudget.PipelineCapacity);
        for (int i = 0; i < VideoFrameBudget.PipelineCapacity; i++)
            Assert.IsTrue(pipeline.TryAcquire(), "a pipeline must not be starved");
        Assert.IsFalse(pipeline.TryAcquire(), "but a runaway is still caught");
    }

    [TestMethod]
    public void ThePeakIsReportedBecauseNothingElseCanSeeIt()
    {
        // The codec's own depth is latency no other measurement reaches, and
        // exposing it is the point of the encoder's budget now that it is not
        // pacing anything.
        var budget = new VideoFrameBudget(8);
        for (int i = 0; i < 5; i++) budget.TryAcquire();
        for (int i = 0; i < 5; i++) budget.Release();

        Assert.AreEqual(0, budget.InFlight);
        Assert.AreEqual(5, budget.HighWater);
        StringAssert.Contains(budget.Summary(), "peak=5");
    }

    [TestMethod]
    public void ConcurrentAcquiresNeverExceedTheCap()
    {
        // The capture thread submits and the MFT's pump thread releases, on
        // different threads. A cap that only holds single-threaded is not a cap.
        var budget = new VideoFrameBudget();
        int highWater = 0;
        object peak = new();

        Parallel.For(0, 8, _ =>
        {
            for (int i = 0; i < 500; i++)
            {
                if (!budget.TryAcquire()) continue;
                lock (peak) highWater = System.Math.Max(highWater, budget.InFlight);
                budget.Release();
            }
        });

        Assert.IsTrue(highWater <= VideoFrameBudget.ProtocolCapacity, $"peaked at {highWater}");
        Assert.AreEqual(0, budget.InFlight);
    }
}
